crates/bevy_asset/src/server/mod.rs

```diff
@@ -58,6 +60,108 @@ pub(crate) struct AssetServerData {
     sources: AssetSources,
     mode: AssetServerMode,
     meta_check: AssetMetaCheck,
+    pub(crate) hooks: RwLock<AssetHooks>,
+}
+
+struct TypeMap(hashbrown::HashMap<TypeId, Box<dyn std::any::Any + Send + Sync>>);
+
+impl TypeMap {
+    pub fn set<T: std::any::Any + Send + Sync + 'static>(&mut self, t: T) {
+        self.0.insert(TypeId::of::<T>(), Box::new(t));
+    }
+    pub fn has<T: 'static + Send + Sync + std::any::Any>(&self) -> bool {
+        self.0.contains_key(&TypeId::of::<T>())
+    }
+
+    pub fn get_mut<T: 'static + Send + Sync + std::any::Any>(&mut self) -> Option<&mut T> {
+        self.0
+            .get_mut(&TypeId::of::<T>())
+            .map(|t| t.downcast_mut::<T>().unwrap())
+    }
+}
+#[allow(clippy::unused_unit)]
+type AssetHookVec<A> =
+    Vec<Box<dyn FnMut(&mut A, &mut bevy_ecs::world::World) -> () + Send + Sync + 'static>>;
+
+pub struct AssetHooks(TypeMap);
+
+impl Default for AssetHooks {
+    fn default() -> Self {
+        Self(TypeMap(hashbrown::HashMap::new()))
+    }
+}
+
+impl<A: Asset> AssetWrapper<A> {
+    fn new(asset: &mut A) -> Self {
+        Self {
+            ptr: asset as *mut A,
+            flag: Default::default(),
+        }
+    }
+}
+/// Safe wrapper around asset pointers to allow for passing into systems.
+pub struct AssetWrapper<A: Asset> {
+    ptr: *mut A,
+    flag: Arc<AtomicBool>,
+}
+
+impl<A: Asset> Deref for AssetWrapper<A> {
+    type Target = A;
+    fn deref(&self) -> &A {
+        // SAFETY: This is safe because we ensure that we flag right after the system that gets the asset
+        // that it has dropped the AssetWrapper. If it hasn't we panic.
+        unsafe { &*self.ptr }
+    }
+}
+
+impl<A: Asset> DerefMut for AssetWrapper<A> {
+    fn deref_mut(&mut self) -> &mut A {
+        // SAFETY: This is safe because we ensure that we flag right after the system that gets the asset
+        // that it has dropped the AssetWrapper. If it hasn't we panic.
+        unsafe { &mut *self.ptr }
+    }
+}
+
+impl<A: Asset> Drop for AssetWrapper<A> {
+    fn drop(&mut self) {
+        self.flag.store(true, Ordering::Release);
+    }
+}
+
+impl AssetHooks {
+    fn add_asset_hook<A, Out, Marker, F>(&mut self, f: F)
+    where
+        A: Asset,
+        F: IntoSystem<AssetWrapper<A>, Out, Marker> + Sync + Send + 'static + Clone,
+    {
+        if !self.0.has::<AssetHookVec<A>>() {
+            self.0.set::<AssetHookVec<A>>(Vec::new());
+        }
+
+        let hooks = self.0.get_mut::<AssetHookVec<A>>().unwrap();
+        hooks.push(Box::new(move |asset, world| {
+            let mut system: F::System = IntoSystem::into_system(f.clone());
+            system.initialize(world);
+            let asset_wrapper = AssetWrapper::new(asset);
```

Comment:
I feel like the initialization logic shouldn't be ran in the closure, but rather in the `trigger` method. This would reduce code bloat.